APIKit/API.swift

```diff
@@ -1,137 +1,92 @@
-        let mainQueue = dispatch_get_main_queue()
-        
-        if let URLRequest = request.URLRequest {
-            let task = URLSession.dataTaskWithRequest(URLRequest)
-            
-            task.request = Box(request)
-            task.completionHandler = { data, URLResponse, connectionError in
-                if let error = connectionError {
-                    dispatch_async(mainQueue) { handler(.failure(error)) }
+        guard let task = URLSession.dataTaskWithRequest(URLRequest) else {
+            notifyError(.InvalidRequest)
+            return nil
+        }
+
+        task.request = Box(request)
+        task.completionHandler = { data, URLResponse, connectionError in
+            if let error = connectionError {
+                notifyError(.ConnectionError(error: error))
+                return
+            }
+
+            guard let HTTPURLResponse = URLResponse as? NSHTTPURLResponse else {
+                notifyError(.UnexpectedResponse)
+                return
+            }
+
+            do {
+                let object = try request.responseBodyParser.parseData(data)
+                if !request.acceptableStatusCodes.contains(HTTPURLResponse.statusCode) {
+                    do {
+                        let error = try request.errorFromObject(object, URLResponse: HTTPURLResponse)
+                        notifyError(.ResponseError(error: error))
+                    } catch {
+                        notifyError(.UnexpectedResponse)
+                    }
                     return
                 }
-                
-                let statusCode = (URLResponse as? NSHTTPURLResponse)?.statusCode ?? 0
-                if !contains(self.acceptableStatusCodes, statusCode) {
-                    let error = self.responseBodyParser.parseData(data).analysis(
-                        ifSuccess: { self.responseErrorFromObject($0) },
-                        ifFailure: { $0 }
-                    )
-
-                    dispatch_async(mainQueue) { handler(.failure(error)) }
-                    return
+
+                let response = try request.responseFromObject(object, URLResponse: HTTPURLResponse)
+                dispatch_async(dispatch_get_main_queue()) {
+                    handler(.success(response))
                 }
-                
-                let mappedResponse: Result<T.Response, NSError> = self.responseBodyParser.parseData(data).flatMap { rawResponse in
-                    if let response = T.responseFromObject(rawResponse) {
-                        return .success(response)
-                    } else {
-                        let userInfo = [NSLocalizedDescriptionKey: "failed to create model object from raw object."]
-                        let error = NSError(domain: APIKitErrorDomain, code: 0, userInfo: userInfo)
-                        return .failure(error)
-                    }
+            } catch {
```

Comment:
This could be written as following:

```swift
} catch let error as APIKitError {
    notifyError(error)
} catch {
    notifyError(.UnexpectedResponse)
}
```